synapse/lib/httpapi.py

```diff
@@ -1240,3 +1243,180 @@ async def get(self):
 
         resp = await self.cell.getCoreInfoV2()
         return self.sendRestRetn(resp)
+
+class ExtApiHandler(StormHandler):
+    '''
+    /api/ext/.*
+    '''
+
+    storm_prefix = 'init { $request = $lib.cortex.httpapi.response($_http_request_info) }'
+
+    async def get(self, path):
+        return await self._runHttpExt('get', path)
+
+    async def head(self, path):
+        return await self._runHttpExt('head', path)
+
+    async def post(self, path):
+        return await self._runHttpExt('post', path)
+
+    async def put(self, path):
+        return await self._runHttpExt('put', path)
+
+    async def delete(self, path):
+        return await self._runHttpExt('delete', path)
+
+    async def patch(self, path):
+        return await self._runHttpExt('patch', path)
+
+    async def options(self, path):
+        return await self._runHttpExt('options', path)
+
+    async def _runHttpExt(self, meth, path):
+        core = self.getCore()
+        adef, args = await core.getHttpExtApiByPath(path)
+        if adef is None:
+            self.set_status(404)
+            self.sendRestErr('NoSuchPath', f'No Extended HTTP API endpoint matches {path}')
+            return await self.finish()
+
+        useriden = adef.get('owner')
+
+        if adef.get('authenticated'):
+
+            requester = await self.useriden()
+
+            if requester is None:
+                await self.reqAuthUser()
+                return
+
+            for pdef in adef.get('perms'):
+                if not await self.allowed(pdef.get('perm'), default=pdef.get('default')):
+                    return
+
+            if adef.get('runas') == 'user':
+                useriden = requester
+
+        # We flatten the request headers and parameters into a flat key/valu map.
+        # The first instance of a given key wins.
```

Comment:
Do we want to return the raw list values for the parameters + headers; and then flatten them on the cortex side upon first access?

That would future proof us from making that raw data available in the multiple parameter / repeated header case.

Reply:
I think we can go with it as is for now since the currently plumbed use case is most likely. We can add plumbing for the raw lists later if we want. :+1: